File: backend/app/models/order.py
"""
Order, OrderItem, and Payment models.
Full order lifecycle with pickup times, cake messages, and Stripe integration.
"""

import enum
import uuid
from datetime import datetime, timezone
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class OrderStatus(str, enum.Enum):
    """Order lifecycle statuses."""
    DRAFT = "draft"                      # Cart-like, not submitted
    PENDING = "pending"                  # Submitted, awaiting payment
    PENDING_APPROVAL = "pending_approval"  # Auth held, awaiting admin approval
    PAID = "paid"                        # Payment confirmed via Stripe webhook
    CONFIRMED = "confirmed"              # Admin confirmed / prep started
    PREPARING = "preparing"              # Being made
    READY = "ready"                      # Ready for pickup
    COMPLETED = "completed"              # Customer picked up
    CANCELLED = "cancelled"              # Cancelled by admin or system
    REFUNDED = "refunded"                # Payment refunded


class PaymentStatus(str, enum.Enum):
    """Payment statuses."""
    PENDING = "pending"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    REFUNDED = "refunded"
    PARTIALLY_REFUNDED = "partially_refunded"


class Order(Base):
    """Order model — represents a customer order."""

    __tablename__ = "orders"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    order_number: Mapped[str] = mapped_column(
        String(20), unique=True, nullable=False, index=True
    )
    customer_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )

    # Status
    status: Mapped[OrderStatus] = mapped_column(
        Enum(OrderStatus), default=OrderStatus.PENDING, nullable=False, index=True
    )

    # Customer info (denormalized for order record)
    customer_name: Mapped[str] = mapped_column(String(255), nullable=False)
    customer_email: Mapped[str] = mapped_column(String(255), nullable=False)
    customer_phone: Mapped[str | None] = mapped_column(String(20), nullable=True)

    # Pickup
    pickup_date: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    pickup_time_slot: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Cake-specific
    cake_message: Mapped[str | None] = mapped_column(String(200), nullable=True)
    has_cake: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False, index=True)
    special_instructions: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Pricing
    subtotal: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False, default=Decimal("0.00"))
    tax_amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False, default=Decimal("0.00"))
    discount_amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False, default=Decimal("0.00"))
    total: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False, default=Decimal("0.00"))

    # Discount / promo
    discount_code: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Admin notes
    admin_notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False, index=True
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    paid_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # Relationships
    items: Mapped[list["OrderItem"]] = relationship(
        back_populates="order", cascade="all, delete-orphan", lazy="selectin"
    )
    payment: Mapped["Payment | None"] = relationship(
        back_populates="order", uselist=False, cascade="all, delete-orphan", lazy="selectin"
    )

    def __repr__(self) -> str:
        return f"<Order {self.order_number} ({self.status.value})>"


class OrderItem(Base):
    """Individual line item in an order."""

    __tablename__ = "order_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("orders.id", ondelete="CASCADE"), nullable=False, index=True
    )
    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id", ondelete="SET NULL"), nullable=True
    )
    variant_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("product_variants.id", ondelete="SET NULL"), nullable=True
    )

    # Denormalized product info (snapshot at time of order)
    product_name: Mapped[str] = mapped_column(String(255), nullable=False)
    variant_name: Mapped[str | None] = mapped_column(String(100), nullable=True)

    # Pricing
    unit_price: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    quantity: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    line_total: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)

    # Cake message for this specific item
    cake_message: Mapped[str | None] = mapped_column(String(200), nullable=True)

    # Relationships
    order: Mapped["Order"] = relationship(back_populates="items")

    def __repr__(self) -> str:
        return f"<OrderItem {self.product_name} x{self.quantity}>"


class Payment(Base):
    """Payment record linked to a Stripe transaction."""

    __tablename__ = "payments"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("orders.id", ondelete="CASCADE"), unique=True, nullable=False
    )

    # Stripe fields
    stripe_checkout_session_id: Mapped[str | None] = mapped_column(
        String(255), unique=True, nullable=True, index=True
    )
    stripe_payment_intent_id: Mapped[str | None] = mapped_column(
        String(255), unique=True, nullable=True, index=True
    )
    stripe_customer_id: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # Payment details
    amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    currency: Mapped[str] = mapped_column(String(3), default="aud", nullable=False)
    status: Mapped[PaymentStatus] = mapped_column(
        Enum(PaymentStatus), default=PaymentStatus.PENDING, nullable=False, index=True
    )
    payment_method: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Refund tracking
    refund_amount: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True)
    refund_reason: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Failure tracking
    failure_code: Mapped[str | None] = mapped_column(String(100), nullable=True)
    failure_message: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Raw webhook data for auditing
    webhook_data: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    order: Mapped["Order"] = relationship(back_populates="payment")

    def __repr__(self) -> str:
        return f"<Payment {self.amount} {self.currency} ({self.status.value})>"

File: admin_frontend/src/types/order.ts
export type OrderStatus =
  | 'draft'
  | 'pending'
  | 'pending_approval'
  | 'paid'
  | 'confirmed'
  | 'preparing'
  | 'ready'
  | 'completed'
  | 'cancelled'
  | 'refunded';

export type PaymentStatus =
  | 'pending'
  | 'succeeded'
  | 'failed'
  | 'refunded'
  | 'partially_refunded';

export interface OrderItemResponse {
  id: string;
  product_id: string | null;
  variant_id: string | null;
  product_name: string;
  variant_name: string | null;
  unit_price: number;
  quantity: number;
  line_total: number;
  cake_message: string | null;
}

export interface PaymentResponse {
  id: string;
  order_id: string;
  stripe_checkout_session_id: string | null;
  stripe_payment_intent_id: string | null;
  amount: number;
  currency: string;
  status: PaymentStatus;
  payment_method: string | null;
  refund_amount: number | null;
  failure_code: string | null;
  failure_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface Order {
  id: string;
  order_number: string;
  customer_id: string | null;
  status: OrderStatus;
  customer_name: string;
  customer_email: string;
  customer_phone: string | null;
  pickup_date: string | null;
  pickup_time_slot: string | null;
  cake_message: string | null;
  has_cake: boolean;
  special_instructions: string | null;
  subtotal: number;
  tax_amount: number;
  discount_amount: number;
  total: number;
  discount_code: string | null;
  admin_notes: string | null;
  items: OrderItemResponse[];
  payment: PaymentResponse | null;
  created_at: string;
  updated_at: string;
  paid_at: string | null;
  completed_at: string | null;
}

export interface OrderListItem {
  id: string;
  order_number: string;
  customer_name: string;
  status: OrderStatus;
  has_cake: boolean;
  total: number;
  pickup_date: string | null;
  created_at: string;
}

export interface OrderItemCreate {
  product_id: string;
  variant_id?: string;
  quantity?: number;
  cake_message?: string;
}

export interface OrderCreate {
  items: OrderItemCreate[];
  customer_name: string;
  customer_email: string;
  customer_phone?: string;
  pickup_date?: string;
  pickup_time_slot?: string;
  cake_message?: string;
  special_instructions?: string;
  discount_code?: string;
}

export interface OrderUpdateAdmin {
  status?: OrderStatus;
  pickup_date?: string;
  pickup_time_slot?: string;
  admin_notes?: string;
}

export interface RefundRequest {
  amount?: number;
  reason: string;
}

// Legacy compatibility types still used by some UI components.
export type PaymentMethod = 1 | 2 | 3 | 4 | 5 | string;

export interface OrderDto {
  id: string;
  product?: string;
  date?: string;
  total: number;
  status?: OrderStatus | number;
  payment_method?: PaymentMethod;
}
